test/test.c

```diff
@@ -32058,6 +32246,920 @@ test_KeyValueMirrorCrossDomains(void)
+        opts = _createReconnectOptions();
+
+    if ((opts == NULL) || 
+        (natsOptions_SetURL(opts, NATS_DEFAULT_URL) != NATS_OK) ||
+        (natsOptions_SetAllowReconnect(opts, false) != NATS_OK))
+    {
+        FAIL("Unable to setup test for MicroConnectionEvents!");
+    }
+
+    serverPid = _startServer("nats://127.0.0.1:4222", NULL, true);
+    CHECK_SERVER_STARTED(serverPid);
+
+    test("Connect for microservice: ");
+    testCond(NATS_OK == natsConnection_Connect(&nc, opts));
+
+    test("Start microservice: ");
+    testCond(NULL == micro_AddService(&m, nc, &cfg));
+
+    test("Test microservice is running: ");
+    testCond(!microService_IsStopped(m))
+
+    test("Stop the server: ");
+    testCond((_stopServer(serverPid), true));
+
+    test("Wait for the service to stop: ");
+    natsMutex_Lock(arg.m);
+    while ((s != NATS_TIMEOUT) && !arg.done)
+        s = natsCondition_TimedWait(arg.c, arg.m, 10000);
+    natsMutex_Unlock(arg.m);
+    testCond(arg.done);
+
+    test("Test microservice is not running: ");
+    testCond(microService_IsStopped(m))
+
+    microService_Destroy(m);
+    natsOptions_Destroy(opts);
+    natsConnection_Destroy(nc);
+    _destroyDefaultThreadArgs(&arg);
+}
+
+void micro_async_error_handler(microService *m, microEndpoint *ep, natsStatus s)
+{
+    struct threadArg *arg = (struct threadArg*) microService_GetState(m);
+
+    natsMutex_Lock(arg->m);
+    arg->status = s;
+    natsCondition_Broadcast(arg->c);
+    natsMutex_Unlock(arg->m);
+}
+
+static void
+test_MicroAsyncErrorHandler(void)
+{
+    natsStatus          s;
+    natsConnection      *nc       = NULL;
+    natsOptions         *opts     = NULL;
+    natsSubscription    *sub      = NULL;
+    natsPid             serverPid = NATS_INVALID_PID;
+    struct threadArg    arg;
+    microService *m = NULL;
+    microServiceConfig cfg = {
+        .Name = "test",
+        .Version = "1.0.0",
+        .DoneHandler = micro_service_done_handler,
+        .ErrHandler = micro_async_error_handler,
+        .State = &arg,
+    };
+
+    s = _createDefaultThreadArgsForCbTests(&arg);
+    if (s != NATS_OK)
+        FAIL("Unable to setup test!");
+
+    arg.status = NATS_OK;
+    arg.control= 7;
+
+    s = natsOptions_Create(&opts);
+    IFOK(s, natsOptions_SetURL(opts, NATS_DEFAULT_URL));
+    IFOK(s, natsOptions_SetMaxPendingMsgs(opts, 10));
+
+    if (s != NATS_OK)
+        FAIL("Unable to create options for test AsyncErrHandler");
+
+    serverPid = _startServer("nats://127.0.0.1:4222", NULL, true);
+    CHECK_SERVER_STARTED(serverPid);
+
+    test("Connect to NATS: ");
+    testCond(NATS_OK == natsConnection_Connect(&nc, opts));
+
+    test("Start microservice: ");
+    testCond(NULL == micro_AddService(&m, nc, &cfg));
+
+    test("Test microservice is running: ");
+    testCond(!microService_IsStopped(m))
+
+    test("Subscribe to async_test: ");
+    testCond(NATS_OK ==  natsConnection_Subscribe(&sub, nc, "async_test", _recvTestString, (void*) &arg));
```

Comment:
With the "arg.control" that you have set, the callback will block after receiving the first message, and therefore would need to be released as such (say before calling `microService_Destroy(m);`):
```
    // Release the subscription
    natsMutex_Lock(arg.m);
    arg.closed = true;
    natsCondition_Broadcast(arg.c);
    natsMutex_Unlock(arg.m);
```

Reply:
@kozlovic Thanks! I kinda got it by looking at the non-micro async error test. Is there a doclet or a comment that documents the control values?